### DISCUSSION TITLE
Layer Annotation

### DISCUSSION THREAD

**Discussion by vsoch** [supercontainers/containers-wg#7] [Ideas]
This is a discussion thread for the [Layer Annotation](https://supercontainers.github.io/containers-wg/ideas/layer-annotation/) idea. You can:

 - vote this discussion up to express interest in the idea
 - add comments, feedback, or ideas for discussion in this thread.

**Comment by dtrudg** [comment]
> When we decompress blobs and plop them into a new container (typically Singularity on HPC) it’s a little bit of a “let’s wait and see” scenario to see if we will have enough space on the filesystem. The sizes on Docker Hub reported the compressed layer sizes, and not the actual contents. Would it be possible to calculate the size of a layer when it’s being created, and then keep it alongside the .tar.gz, or even in a header somewhere?

I think this is something where we are only ever going to be able to give an upper bound for extracted size, not a truly accurate figure. Knowing what's in a layer still doesn't let us understand the extracted disk usage for a container with many layers. Consider the HPC runtimes that don't preserve and overlay layers, but extract layers on top of eachother into a flattened directory:

- Layer 3 extracts a 30MB file /bin/myprog
- Layer 5 replaces /bin/myprog with a 35MB file

The layer-by-layer usage would suggest a total of 65MB, but in reality only 35MB is consumed. The CharlieCloud issue linked in the idea, and the docs it refers to indicate the issues nicely: https://github.com/hpc/charliecloud/commit/234b18b41d99cfdfb0e1ea8b53d47ac0221cca92

Getting an accurate total 'disk space needed' is also going to be quite runtime dependent.

For Singularity a straight pull is going to...

- Pull the (missing) OCI layers to the cache directory (usually in $HOME)... we can work this usage out.
- Extract the layers into a temporary directory (see above).. this will be deleted on completion, and is usually in $TMPDIR.
- Create a SIF in which the container content is SquashFS compressed.. the compression will differ from the layer compressed sizes and can't be predicted in advance.

Other runtimes like CharlieCloud, Enroot, which do or can build to SquashFS or flattened tar are going to have similar but different disk usage concerns here, where the total disk usage probably can't be predicted even if we know the size of the layer content.
